### PR TITLE
Restrict access to pages based on cookies/form submissions

### PR DESCRIPTION
**Pull request by joshdarby**
## Changes

This pull request makes the following changes:

- Adds a new checkbox metabox to the editor for posts and pages to toggle whether or not the content should be restricted behind a cookie.
- Adds a new `set_cookie_on_form_submission` function that sets a cookie for 365 days in the users browser once they submit a form
- Adds a new `restrict_access_to_pages_by_cookie` function that restricts access to a page if the aforementioned checkbox is checked. However, if the user has a cookie set named `unrestrict_POST_ID` (POST_ID will be an actual post id) and it's value is set to true, the page will show.
    - If user lands on a restricted page and doesn't have a cookie set, the form will show. Once filled out, user is redirected to the page with content shown.
    - If user is on a restricted child page, it looks to see if a cookie is set to grant access to its top level parent page. If it is, access is granted
     - If a user is on a restricted child page and no top level parent cookie is set, it redirects the user to the top level parent page to fill out the form.

Notes:
- The name of the form **must** match the top level parent page that is restricted. 
- In order to know what post the form grants access to, a hidden field **must** be placed as the **last** field in the form and the value needs to be the post ID of the post to grant access to

Restrict access metabox:
Screen Shot 2020-05-12 at 1 54 57 PM

Top level page with form:
ezgif-3-3ecde6edf709

Page with no matching form but that is restricted:
Screen Shot 2020-05-12 at 1 54 37 PM

## Why

For #161

## Testing/Questions

Features that this PR affects:

- Pages and posts that enable the `Restrict Access` checkbox

Questions that need to be answered before merging:

- [ ] Is this PR targeting the correct branch in this repository?
- [ ] Any logic I missed?
- [ ] Any other edge cases we need to watch out for?

Steps to test this PR:

1. Create a new page or pick a page to test with
2. Enable the "Restrict Access" checkbox
3. Create a new GF form
    - make the title of the form the exact title of your test page
    - make the last field of the form a hidden field and give it the value of your test page
4. Add a child page to your test page and restrict it.
5. Add a child page to your child page and restrict it.
6.  Open your test page in an incognito window and fill out the form. Hopefully you're redirected to the page with the content now and are able to also view the child pages.

### CODE REVIEW

**Comment by benlk** [comment]
Two implementation-detail questions:
- How does this work if WPE has cached the page HTML? Does WPE in production environments serve different page results with and without the cookie?
- Is it more future-proof if the form name is the ID of the post instead of the name of the post? That way, folks could rename the post without breaking the form.

File: wp-content/themes/cjet/functions.php [hunk 1]
```diff
@@ -113,3 +113,139 @@ function load_dashicons_front_end() {
...
+
+			$ancestors = get_post_ancestors( $post->ID );
+			$root = count( $ancestors ) - 1;
+			$parent = $ancestors[$root];
+
+		} else {
+
+			$parent = $post->ID;
+
+		}
```
[the remark this answers is not in the document]
Comment (joshdarby) [reply wp-content/themes/cjet/functions.php:202]: We could do that. In your opinion, what would make that more beneficial than this way?

**Comment by joshdarby** [comment]
> Two implementation-detail questions:
> 
> * How does this work if WPE has cached the page HTML? Does WPE in production environments serve different page results with and without the cookie?

That's a great question that I'm not sure about until we put it up on a WPE environment and test it out. We may possibly need to add some cachebusting into it.

> * Is it more future-proof if the form name is the ID of the post instead of the name of the post? That way, folks could rename the post without breaking the form.

I'm torn between yes and no here. Yes, the ID would be better in a lot of ways. But then we'll have users looking through x amount of forms trying to find the correct one because they don't know the actual ID of the page/post form they're looking for, and I'm not sure if it's a reasonable ask to expect them to know the ID.

File: wp-content/themes/cjet/functions.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (benlk) [reply wp-content/themes/cjet/functions.php:202]: The possibility of a metered page having a grandchild or great-grandchild page. They're already using great-grandchild pages: https://learn.inn.org/startup-introduction/chapter-1-where-to-start-in-nonprofit-news/creating-a-business-model/idea-generation/

**Comment by benlk** [comment]
> But then we'll have users looking through x amount of forms trying to find the correct one because they don't know the actual ID of the page/post form they're looking for, and I'm not sure if it's a reasonable ask to expect them to know the ID.

We could add to the meta box a link to the appropriate form. In pseudocode:

```
$form_exists = false;
for post_id in array( this post ID, ancestor post IDs ) {
    if form_exists( $post_id ) {
        $form_exists = true;
        "This post's visibility would be controlled by <this form>"
    }
}
if ( ! $form_exists ) {
   no form exists for this post. [create one](/wp-admin/admin.php?page=gf_new_form) with the title $post_id
}
```

There doesn't appear to be a way to prefill the form title in the create-new-form form, unfortunately. I was hoping `&new_form_title=12345` or `&title=12345` would prefill it with "12345", but the plugin contains no logic to do that:

https://github.com/wp-premium/gravityforms/blob/ca668e0dc28bc42a8634be619837bbdc8f445d16/form_list.php#L96

**Comment by joshdarby** [comment]
> > But then we'll have users looking through x amount of forms trying to find the correct one because they don't know the actual ID of the page/post form they're looking for, and I'm not sure if it's a reasonable ask to expect them to know the ID.
> 
> We could add to the meta box a link to the appropriate form. In pseudocode:
> 
> ```
> $form_exists = false;
> for post_id in array( this post ID, ancestor post IDs ) {
>     if form_exists( $post_id ) {
>         $form_exists = true;
>         "This post's visibility would be controlled by <this form>"
>     }
> }
> if ( ! $form_exists ) {
>    "no form exists for this post. [create one](/wp-admin/admin.php?page=gf_new_form) with the title $post_id"
> }
> ```
> 
> There doesn't appear to be a way to prefill the form title in the create-new-form form, unfortunately. I was hoping `&new_form_title=12345` or `&title=12345` would prefill it with "12345", but the plugin contains no logic to do that:
> 
> https://github.com/wp-premium/gravityforms/blob/ca668e0dc28bc42a8634be619837bbdc8f445d16/form_list.php#L96

@benlk I like this idea, but the inability to populate the form title with the post ID makes me weary because then we're back at the question of whether or not its a reasonable ask to expect the user to know how to find the post ID :/

File: wp-content/themes/cjet/functions.php [hunk 2]
```diff
@@ -113,3 +113,139 @@ function load_dashicons_front_end() {
...
+
+			}
+
+		}
+
+		// make sure the GF methods exist before we try using them
+		// that way we don't see any scary 500 errors
+		if( method_exists( 'RGFormsModel', 'get_form_id' ) && method_exists( 'RGForms', 'get_form' ) ) {
+
+			// try and find a form that matches the current post title
```
Comment (benlk) [inline wp-content/themes/cjet/functions.php:227]: Does this need a fallback case for if the parent is the page with the form?

Comment (joshdarby) [reply wp-content/themes/cjet/functions.php:227]: @benlk it shouldn't because a child page shouldn't ever get to this point based on the conditional above that will redirect them to the parent page if a cookie isn't set.

Comment (benlk) [reply wp-content/themes/cjet/functions.php:227]: Oh! I see that now.
